ghiscoding/aurelia-slickgrid
# BREAKING CHANGE: use Slickgrid-Universal monorepo

### Next Major Version `3.x` (WIP)

This will be, for the most part, deleting a lot of code and instead use it from the [Slickgrid-Universal](https://github.com/ghiscoding/slickgrid-universal) lib to have 1 common/core lib. Also that lib was built as a monorepo and with that we'll be able to decouple a few Aurelia-Slickgrid Services (Export to File/Excel, OData, GraphQL). So you will end up with a much smaller bundle since you will be bundling only what you really use (without having for example OData/GraphQL that you might never use). For even more info, you can read a full explanation that was posted under an older release [here](https://github.com/ghiscoding/aurelia-slickgrid/releases/tag/v2.19.1) under the section labeled "What's coming in the Future (much later in the year)"

On the migration side, it should be fairly simple, the only changes will be in regards to the Services that get decoupled, you can take a sneak peak at the changes in the upcoming [Migration Guide](/ghiscoding/aurelia-slickgrid/wiki/Migration-to-3.x)

ETA 🤞 - Wrapping it up for Christmas would be nice 🎁 🎄 🎅

###  [Migration Guide](/ghiscoding/aurelia-slickgrid/wiki/Migration-to-3.x)

#### Minimum Requirement Changes
- TypeScript `>= 4.1.2`
- Aurelia-I18N `>= 3.0.0`
- Aurelia-CLI `>= 2.0.0`
- Bootstrap `>= 4.5.0` (or any other UI framework like Bulma)

#### List of TODOs
##### _if you want to help, just tell me which number you are planning to provide help with_ 
- [x] 1. replace all Formatters with Slickgrid-Universal
- [x] 2. replace all Grouping Formatters with Slickgrid-Universal
- [x] 3. replace all Aggregators with Slickgrid-Universal
- [x] 4. replace all Editors with Slickgrid-Universal
- [x] 5. replace all Editor Valitors with Slickgrid-Universal
- [x] 6. replace all Filters with Slickgrid-Universal
   - (requires DI) extend `AutoCompleteFilter` and `SelectFilter` so that we can watch collection changes with BindingEngine
   - goal change, I might be able to implement collection watch directly in Slickgrid-Universal
- [x] 7. replace all Filter Conditions with Slickgrid-Universal
- [x] 8. replace all Models with Slickgrid-Universal (enums, interfaces)
- [x] 9. replace all Sort Comparers with Slickgrid-Universal
- [x] 10. replace all Services with Slickgrid-Universal
   - (requires DI) need to extend `FilterService` for the 2x Filters mentioned at step 6) to be extended
   - goal change (same as 6.), I might be able to implement collection watch directly in Slickgrid-Universal
- [x] 11. replace all Extensions with Slickgrid-Universal
   - [x] keep RowDetail Extension since it has specific Aurelia code
- [x] 12. replace all SASS Stylings with Slickgrid-Universal
- [x] 13. replace all Export to File Service with Slickgrid-Universal Export to File monorepo package
- [x] 14. replace all Export to Excel Service with Slickgrid-Universal Export to Excel monorepo package
- [x] 15. replace all Export to OData Service with Slickgrid-Universal OData monorepo package
- [x] 16. replace all Export to GraphQL Service with Slickgrid-Universal GraphQL monorepo package
- [x] 17. some of the events might not work on the Aurelia-Slickgrid custom element because the services are now in separate monorepo packages (`onGridBeforeExportToFile`, `onGridAfterExportToFile`, `onGridBeforeExportToExcel`, `onGridAfterExportToExcel`)
   - the names might be different, it should be `onBeforeExportToTextFile`, ...
- [x] 18. add a Translate Service wrapper of Aurelia-I18N for Slickgrid-Universal to work properly
- [x] 19. add a PubSub Service wrapper of Aurelia-Event-Aggregator for Slickgrid-Universal to work properly
- [x] 20. upgrade to latest Aurelia-I18N version
- [x] 21. add demo to cover PR #442 and close issue #441
- [x] 22. use Bootstrap 4 as a base instead of BS3
- [x] 23. drop internal copy of 3rd party lib [multiple-select.js](https://github.com/ghiscoding/aurelia-slickgrid/tree/master/src/assets/lib/multiple-select) and use [multiple-select-modified](https://github.com/ghiscoding/multiple-select-modified) instead
- [x] 24. add `yarn.lock` file into Git and make use of [dependabot](https://dependabot.com/), see [dependabot.yml](https://github.com/ghiscoding/slickgrid-universal/blob/master/.github/dependabot.yml) config from Slickgrid-Universal 
- [x] 25. create/update [Migration Guide](/ghiscoding/aurelia-slickgrid/wiki/Migration-to-3.x) Wiki to next version `3.x`
- [ ] 26. update all necessary Wikis
   - Export to Excel from `@slickgrid-universal/excel-export`
   - Export to Text File from `@slickgrid-universal/text-export`
   - OData Backend Service from `@slickgrid-universal/odata`
   - GraphQL Backend Service from `@slickgrid-universal/graphql`
- [x] 27. rename/move `master` branch into `version2` branch to keep ref to older version
- [x] 28. create a Git Tag on Aurelia-Slickgrid-Demos of `2.x` version before rewriting to newer version
- [ ] 29. add the new SlickGrid Composite Editor + Example
- [x] 30. Convert dynamic loading of SlickGrid plugins/controls from `require(plugin)` to ~`await import(plugin)`~
    - this is causing way too many racing conditions, I will take a look at that much later in the future in separate version.
    - see this [article](https://mariusschulz.com/blog/dynamic-import-expressions-in-typescript) or this other [article](https://blog.mgechev.com/2019/05/11/dynamic-imports-javascript/)
- [x] 31. jQueryUI is imported separately in Slickgrid-Universal (via `main.ts`), check to see if we need to do that here
- [x] 32. search for any leftover of `TODO`, `@ts-ignore` and/or `@deprecated` code
- [x] 33. all Cypress E2E tests should work
- [x] 34. all Jest Unit Tests should work and try to keep a 100% test coverage
- [x] 35. remove a few of the npm scripts, we probably don't need to bundle anymore for (`jspm`, `system`)
- [x] 36. rename monorepo `FileExportService` to `TextExportService` and also couple of option changes:
   - deprecate `exportOptions` use new option `exportTextOptions`
   - deprecate `enableExport` use new option `enableTextExport`
- [x] 37. remove prefix `asg` and `sg` from all events in the global grid options (user can put it back in their own configs).
